### PR TITLE
ENT-6018 Teach tom to add packages to package_mapping.json file

### PR DESCRIPTION
**Pull request by Lex-2008**
Sample PR: https://github.com/cfengine/system-testing/pull/333

----

#

### CODE REVIEW

File: tom/packages.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import re
+import json
+import requests
+import collections
+import datetime
+import hashlib
+import urllib.request
+import logging as log
```
Comment (vpodzime) [inline tom/packages.py:8]: The 3 above imports are unused.

File: tom/packages.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from tom.git import GitRepo
+from tom.utils import pretty
```
Comment (vpodzime) [inline tom/packages.py:10]: And so is this one.

File: tom/packages.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        repo_name = 'system-testing'
+        upstream_name = 'cfengine'
+        local_path = "../" + repo_name
+        repo = GitRepo(local_path, repo_name, upstream_name, self.username, "master")
+        timestamp = re.sub('[^0-9-]', '_', str(datetime.datetime.today()))
+        new_branchname = 'packages_mapping-{}'.format(timestamp)
+        repo.checkout(new_branchname, new=True)
+        # load current file
+        packages_mapping_file_path = 'deployment_tests/packages_mapping.json'
+        packages_mapping = json.loads(repo.get_file(packages_mapping_file_path), object_pairs_hook=collections.OrderedDict)
```
Comment (vpodzime) [inline tom/packages.py:40]: Please split this into multiple lines with exceptions being handled gracefully.

File: tom/packages.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            for product in ['community','enterprise']:
+                releases_data = requests.get('https://cfengine.com/release-data/%s/releases.json' % product).json()
+                release_url, version = next((release['URL'], release['version']) for release in releases_data['releases']
+                                            if "lts_branch" in release and release["lts_branch"] == branch
+                                            and "latest_on_branch" in release and release["latest_on_branch"])
+                packages = self.collect_packages(release_url)
+                # sort packages into agent/hub ones
+                if product == 'community':
+                    # for community, they are the same
+                    hub_packages = agent_packages = packages
```
Comment (vpodzime) [inline tom/packages.py:62]: I don't like this double assignment, but YMMV.

File: tom/packages.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                else:
+                    hub_packages = {platform:value for (platform,value) in packages.items() if '_HUB_' in platform}
```
Comment (vpodzime) [inline tom/packages.py:64]: Missing spaces around `:` and `,`.

File: tom/packages.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        #       "agent": {
+        #         "community": {
+        #           "PACKAGES_i386_linux_debian_4": {
+        #             "url": "https://cfengine-package-repos.s3.amazonaws.com/community_binaries/cfengine-community_3.6.6-1_i386.deb"
+        #           },
+        for branch in branches.split(','):
+            self.slack.reply("Updating packages mapping for " + branch)
+            result = {'agent':{},'hub':{}}
+            for product in ['community','enterprise']:
+                releases_data = requests.get('https://cfengine.com/release-data/%s/releases.json' % product).json()
```
Comment (vpodzime) [inline tom/packages.py:54]: Same here with multiple separate steps and exceptions being handled.

File: tom/packages.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        self.slack.reply(pr_text, True)
+
+    def collect_packages(self, url):
+        """Given a release URL, returns a dict where keys are platform names, and values are {'url': 'http...'} """
+        release_data = requests.get(url).json()
+        # release_data['artifacts'] is a dictionary (called 'table'), each element is a list
+        # to flattern it:
+        # [item for table in release_data['artifacts'] for item in release_data['artifacts'][table]]
+        # to choose only items with defined platforms (they are called 'package' and might be undefined)
+        return {item['package']: {'url': item['URL']}
```
Comment (vpodzime) [inline tom/packages.py:94]: A bonus here, although maybe unnecessary, would be to try to get 1 byte from the URL and check the status code to test that the URL is correct.

File: tom/packages.py (hunk 3 above)
Comment (Lex-2008) [reply tom/packages.py:40]: exceptions if there is no `packages_mapping.json` file and if it contains invalid json? okay...

Comment (vpodzime) [reply tom/packages.py:40]: Not a big deal in this case, I originally confused it with the step that fetches and parses the `releases.json` file and stuff. But still nice to at least have the `repo.get_file()` and `json.loads()` done in two steps on two lines, I think.

File: tom/packages.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        repo = GitRepo(local_path, repo_name, upstream_name, self.username, "master")
+        timestamp = re.sub('[^0-9-]', '_', str(datetime.datetime.today()))
+        new_branchname = 'packages_mapping-{}'.format(timestamp)
+        repo.checkout(new_branchname, new=True)
+        # load current file
+        packages_mapping_file_path = 'deployment_tests/packages_mapping.json'
+        try:
+            packages_mapping_contents = repo.get_file(packages_mapping_file_path)
+        except FileNotFoundError:
+            raise RepoFileNotFoundException('file %s not found in repo. Is repo broken?' % packages_mapping_file_path)
```
Comment (vpodzime) [inline tom/packages.py:59]: A good candidate from `raise ... from ...` I think.

Comment (vpodzime) [reply tom/packages.py:59]: And the same applies to other cases below.

File: tom/packages.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            for (product, codename) in [('community', 'community'), ('enterprise', 'nova')]:
+                releases_url = 'https://cfengine.com/release-data/%s/releases.json' % product
+                releases_request = requests.get(releases_url)
+                if not releases_request.ok:
+                    raise URLDownloadFailureException('failed to download %s, return code %d' % (releases_url, releases_request.status_code))
+                try:
+                    releases_data = releases_request.json()
+                except json.decoder.JSONDecodeError:
+                    raise JSONParsingError('file %s is not a valid JSON' % releases_url)
+                if not 'releases' in releases_data:
```
Comment (vpodzime) [inline tom/packages.py:86]: `'releases' not in releases_data`

File: tom/packages.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                    hub_packages = agent_packages = packages
+                else:
+                    hub_packages = {platform: value for (platform, value) in packages.items() if '_HUB_' in platform}
+                    agent_packages = {platform: value for (platform, value) in packages.items() if '_HUB_' not in platform}
+                result['agent'][codename] = agent_packages
+                result['hub'][codename] = hub_packages
+            packages_mapping['packages'][version] = result
+            repo.put_file(packages_mapping_file_path, json.dumps(packages_mapping, indent=2))
+            # update version in upgrade_from.json
+            branch_nox=re.sub('\\.x$', '', branch)
```
Comment (vpodzime) [inline tom/packages.py:107]: missing spaces

File: tom/packages.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            upgrade_from_file_path = 'deployment_tests/upgrade_from.json'
+            upgrade_from = json.loads(repo.get_file(upgrade_from_file_path), object_pairs_hook=collections.OrderedDict)
```
Comment (vpodzime) [inline tom/packages.py:109]: No exception handling here.

File: tom/packages.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    def collect_packages(self, url):
+        """Given a release URL, returns a dict where keys are platform names, and values are {'url': 'http...'} """
+        release_request = requests.get(url)
+        if not release_request.ok:
+            raise URLDownloadFailureException('failed to download %s, return code %d' % (url, release_request.status_code))
+        try:
+            release_data = release_request.json()
+        except json.decoder.JSONDecodeError:
+            raise JSONParsingError('file %s is not a valid JSON' % url)
+        if not 'artifacts' in release_data:
```
Comment (vpodzime) [inline tom/packages.py:132]: `not in`

File: tom/packages.py (hunk 7 above)
Comment (Lex-2008) [reply tom/packages.py:138]: I agree, but probably it should be done by some other script - for example, the one which checks SHAs.

**Comment by Lex-2008** [comment]
commit amended to reflect the fact that masterfiles package must be among hub packages, not agent ones.

**Comment by Lex-2008** [comment]
PR updated to accept also version, in addition to LTS branch name, so Tom can also add non-LRS 3.16.0 packages to package_mapping file! PR: https://github.com/cfengine/system-testing/pull/346

File: tom/packages.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    """Exception that is risen when a JSON file is not a valid JSON"""
+    pass
+
+class JSONStructureError(PackageMapperException):
+    """Exception that is risen when JSON doesn't have expected fields"""
+    pass
+
+def is_branch_or_version(string):
+    '''Tries to figure out if passed argument is branch or version.
+    Returns 'branch', 'version', or False if deduction failed.
```
Comment (vpodzime) [inline tom/packages.py:30]: it actually returns `None` in case of failure

File: tom/packages.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    Branch is either 'master' or something like 3.12.x;
+    version is something like 3.12.5,
+    optionally followed by letter (3.12.5b) for aplha/beta/gamma...zeta,
+    optionally followed by release (3.12.5-2).
+    '''
+    if string == 'master' or re.match('3\.\\d+\.x$', string):
+        return 'branch'
+    if re.match('3\\.\\d+\\.\\d+[a-z]?(-\\d+)?$', string):
+        return 'version'
+    return None
```
Comment (vpodzime) [inline tom/packages.py:40]: Should this maybe raise an exception instead? We wouldn't need the `assert` then.

**Pull request merged**